core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -315,8 +315,8 @@ class half_t {
     // Use non-volatile val_ref to suppress:
     // "warning: implicit dereference will not access object of type ‘volatile
     // __half’ in statement"
-    auto val_ref = const_cast<impl_type&>(val);
-    val_ref      = __float2half(__half2float(const_cast<impl_type&>(val)) +
+    auto& val_ref = const_cast<impl_type&>(val);
+    val_ref       = __float2half(__half2float(const_cast<impl_type&>(val)) +
                            __half2float(rhs.val));
```

Comment:
omg this operator was simply wrong before. val_ref was a value type wasn't it? Not a reference type? Can we get a test for this?

Reply:
Does not belong to this PR IMO

Reply:
Moved to #4198 and asked @e10harvey for a test.

Reply:
These tests are part of #4101 which is ready for another round of reviews. See https://github.com/kokkos/kokkos/pull/4101/files#diff-6980b46765a569d9ae64dffdbe6a0db53b9ed49c8b9157236055a90add5b47afR277-R352.